File: clojure/src/circle_db/layer_update.clj
(ns circle-db.layer-update
  (:require [clojure.set :as set]
            [circle-db.constructs :as c]
            [circle-db.eavt :as e]
            [circle-db.avet :as av]
            [circle-db.veat :as ve]
            [circle-db.vaet :as vae]))

(defn- iter-vals [value]
  (if (set? value) value [value]))

(defn- apply-op [old-val new-val op cardinality]
  (if (= cardinality :db/single)
    new-val
    (case op
      :reset  new-val
      :add    (set/union old-val new-val)
      :remove (set/difference old-val new-val))))

(defn update-entity-in-layer [layer entity-id attr-name new-val op curr-time]
  (if-let [entity (get (:storage layer) entity-id)]
    (let [old-attr  (get-in entity [:attrs attr-name])
          new-value (apply-op (:value old-attr) new-val op (:cardinality old-attr))
          new-attr  (assoc old-attr :value new-value :ts curr-time :prev-ts (:ts old-attr))
          new-entity (assoc-in entity [:attrs attr-name] new-attr)
          remove-from-indexes (fn [l v]
                                (let [d (c/->Datom entity-id attr-name v)]
                                  (-> l
                                      (update :avet av/index-remove d)
                                      (update :veat ve/index-remove d)
                                      (cond-> (= (:type old-attr) :db/ref)
                                        (update :vaet vae/index-remove d)))))
          add-to-indexes (fn [l v]
                           (let [d (c/->Datom entity-id attr-name v)]
                             (-> l
                                 (update :avet av/index-add d)
                                 (update :veat ve/index-add d)
                                 (cond-> (= (:type old-attr) :db/ref)
                                   (update :vaet vae/index-add d)))))]
      (-> layer
          (assoc-in [:storage entity-id] new-entity)
          (update :eavt e/index-add (c/->Datom entity-id attr-name new-value))
          (as-> l (reduce remove-from-indexes l (iter-vals (:value old-attr))))
          (as-> l (reduce add-to-indexes      l (iter-vals new-value)))))
    layer))

File: python/src/layer_update.py
from dataclasses import replace
from constructs import Datom
from eavt import index_add as eavt_add
from avet import index_remove as avet_remove, index_add as avet_add
from veat import index_remove as veat_remove, index_add as veat_add
from vaet import index_remove as vaet_remove, index_add as vaet_add


def _apply_op(old_value, new_val, op, cardinality):
    if cardinality == ":db/single":
        return new_val
    if op == "reset":
        return new_val
    if op == "add":
        return (old_value or set()) | new_val
    if op == "remove":
        return (old_value or set()) - new_val


def update_entity_in_layer(layer, entity_id, attr_name, new_val, op, curr_time):
    entity = layer.storage.get(entity_id)
    if entity is None:
        return layer
    old_attr = entity.attrs[attr_name]
    new_value = _apply_op(old_attr.value, new_val, op, old_attr.cardinality)
    new_attr = replace(old_attr, value=new_value, ts=curr_time, prev_ts=old_attr.ts)
    new_entity = replace(entity, attrs={**entity.attrs, attr_name: new_attr})
    new_storage = {**layer.storage, entity_id: new_entity}

    def iter_vals(v):
        return v if isinstance(v, set) else [v]

    new_eavt = eavt_add(layer.eavt, Datom(entity_id=entity_id, attr_name=attr_name, value=new_value))
    new_avet, new_veat, new_vaet = layer.avet, layer.veat, layer.vaet
    for v in iter_vals(old_attr.value):
        d = Datom(entity_id=entity_id, attr_name=attr_name, value=v)
        new_avet = avet_remove(new_avet, d)
        new_veat = veat_remove(new_veat, d)
        if old_attr.type == ":db/ref":
            new_vaet = vaet_remove(new_vaet, d)
    for v in iter_vals(new_value):
        d = Datom(entity_id=entity_id, attr_name=attr_name, value=v)
        new_avet = avet_add(new_avet, d)
        new_veat = veat_add(new_veat, d)
        if old_attr.type == ":db/ref":
            new_vaet = vaet_add(new_vaet, d)

    return replace(layer, storage=new_storage, eavt=new_eavt, avet=new_avet, veat=new_veat, vaet=new_vaet)
